### PR TITLE
Implement proof of concept scrubber for png and jpeg files

### PR DESCRIPTION
**Pull request by max-b**
This is just a spike/proof of concept for removing metadata from png files. It takes a _slightly_ agressive approach by removing all `eXIf`, `tEXt`, `iTXt`, `zTXt`chunk types. I think it would behoove doing more comprehensive testing on a variety of files and ensuring they maintain readability.

There are still some hairy corners/`TODO`s, but I wanted to get this in front of other people's eyes ASAP.

Thanks!!

### CODE REVIEW

**Comment by max-b** [comment]
Implementing a proof of concept for jpegs ended up being straightforward enough so I just added it here. There are an unfortunate number of code/logic repetitions between the png and jpeg scrubber, but I'm not totally sure there's a sensible way to abstract/DRY them...

File: cli/main.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package main
```
Comment (hwh33) [inline cli/main.go]: This is nit-picky, but Go will kind of assume the name of the command is the folder the main package resides in.  So Go will think of this as a command called "cli".  You can see this when you try to build it without specifying the output, e.g. `go build ./cli`.  You'll get a binary called "cli".

A common pattern is to create a folder called "cmd" and put utilities in subdirectories by name.  So you could have `./cmd/meta-scrubber/main.go`.  Then `go build ./cmd/meta-scrubber` will give you an appropriately named binary.

Just as an FYI in the same vein, if you wanted to add package-level documentation here, govet will tolerate something like
```
// Command meta-scrubber is a cool CLI utility.
package main
```
if the parent directory is named meta-scrubber.

Comment (hwh33) [reply cli/main.go]: Oh it's the linter (`golint`) that inspects doc comments.  I kind of think of the vet and lint tools as the same thing ¯\\_(ツ)_/¯

File: scrubber.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
+// Scrubber provides a streaming metadata remover
+package metascrubber
```
Comment (hwh33) [inline scrubber.go:2]: Go wants package doc to be of the form "Package metascrubber provides a streaming metadata remover."

File: scrubber.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import (
+	"bytes"
+	"io"
+	"net/http"
+
+	"github.com/getlantern/meta-scrubber/jpeg"
+	"github.com/getlantern/meta-scrubber/png"
+)
+
+func GetScrubber(reader io.Reader) (io.Reader, error) {
```
Comment (hwh33) [inline scrubber.go:13]: Everything exported should have a doc comment or `golint` will complain.

File: scrubber.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	// net/http.sniffLen = 512
+	contentTypeHead := make([]byte, 512)
+	n, err := io.ReadFull(reader, contentTypeHead)
+
+	if err != nil {
+		if err == io.EOF {
+			// TODO: what's the best way to return an empty reader
+			return bytes.NewReader([]byte{}), nil
```
Comment (hwh33) [inline scrubber.go]: Re: the best way to return an empty reader, I think what you've got is good.

File: scrubber.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	"github.com/getlantern/meta-scrubber/png"
+)
+
+func GetScrubber(reader io.Reader) (io.Reader, error) {
+	// net/http.sniffLen = 512
+	contentTypeHead := make([]byte, 512)
+	n, err := io.ReadFull(reader, contentTypeHead)
+
+	if err != nil {
+		if err == io.EOF {
```
Comment (hwh33) [inline scrubber.go]: The canonical way to do this is `errors.Is(err, io.EOF)`.  This is relatively new API, so you'll still see lots of `err == io.EOF` around.  It'll probably work in most cases, but it's better to adopt the new way I guess.

[More info here](https://blog.golang.org/go1.13-errors)

Comment (hwh33) [reply scrubber.go]: (same goes for the `io.ErrUnexpectedEOF` check below)

File: scrubber.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+			// TODO: what's the best way to return an empty reader
+			return bytes.NewReader([]byte{}), nil
+		} else if err == io.ErrUnexpectedEOF {
+			return bytes.NewReader(contentTypeHead[:n]), nil
+		} else {
+			return nil, err
+		}
+	}
```
Comment (hwh33) [inline scrubber.go:23]: If you're interested, it looks like this block could be simplified to:
```
if err != nil {
    if errors.Is(err, io.EOF) || errors.Is(err, io.UnexpectedEOF) {
        return bytes.NewReader(contentTypeHead[:n]), nil
    }
    return nil, err
}
```

File: scrubber_test.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+	"bytes"
+	"io/ioutil"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func loadBytes(t *testing.T, name string) []byte {
+	path := filepath.Join("./", "fixtures", name)
```
Comment (hwh33) [inline scrubber_test.go]: I think this won't work on Windows because of the forward-slash right?

Test data in Go generally goes in a folder called "testdata".  There's some support for this in the Go tool, though I don't actually know what that support is... haha.  But yeah, stuff in there should be reference-able via `filepath.Join("testdata", "name")`.

File: scrubber_test.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+func TestScrubbingEmptyReader(t *testing.T) {
+	emptyReader := bytes.NewReader([]byte{})
+	emptyScrubber, err := GetScrubber(emptyReader)
+	if assert.NoError(t, err) {
+		actuallyRead, err := ioutil.ReadAll(emptyScrubber)
+		if assert.NoError(t, err) {
+			assert.Empty(t, actuallyRead)
+		}
+	}
```
Comment (hwh33) [inline scrubber_test.go]: I like `testify/require` for checks that need to pass for the test to proceed.  So this could become:
```
emptyReader := bytes.NewReader([]byte{})
emptyScrubber, err := GetScrubber(emptyReader)
require.NoError(t, err)
actuallyRead, err := ioutil.ReadAll(emptyScrubber)
require.NoError(t, err)
require.Empty(t, actuallyRead)
```

File: scrubber_test.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+	compareImages(t, "kitten-without-meta.jpeg", "kitten-without-meta.jpeg")
+}
+
+func TestScrubbingJpegWithXmpDescription(t *testing.T) {
+	compareImages(t, "kitten-with-xmp-description.jpeg", "kitten-without-meta.jpeg")
+}
+
+func TestScrubbingJpegWithExifMake(t *testing.T) {
+	compareImages(t, "kitten-with-exif-make.jpeg", "kitten-without-meta.jpeg")
+}
```
Comment (hwh33) [inline scrubber_test.go:79]: I feel like these tests belong in the `png` and `jpeg` packages since they're really testing functionality implemented there.  I guess they're kind of integration tests, but there's not much integration logic to test really.

File: jpeg/jpeg.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+import (
+	"bytes"
+	"encoding/binary"
+	"io"
+	"io/ioutil"
+
+	jpegimagestructure "github.com/dsoprea/go-jpeg-image-structure"
+)
+
+var headerBytes []byte = []byte{0xff, 0xd8}
```
Comment (hwh33) [inline jpeg/jpeg.go]: Looks like we only use the length of this.  Maybe `const headerBytes = 2` would be more appropriate?

File: jpeg/jpeg.go [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+	}
+}
+
+func isMetadataType(segmentMarker []byte) bool {
+	// TODO: consider other APP types
+	return segmentMarker[0] == 0xff &&
+		(segmentMarker[1] == jpegimagestructure.MARKER_APP1 || segmentMarker[1] == jpegimagestructure.MARKER_APP0)
+}
+
+func (self *JpegScrubber) Read(buf []byte) (n int, err error) {
```
Comment (hwh33) [inline jpeg/jpeg.go]: Receivers in Go are not typically named `self` or `this` ([here's a style guide of sorts which mentions this](https://github.com/golang/go/wiki/CodeReviewComments#receiver-names)).  It's kind of Go dogmatism, but there are some justifications I guess.  [Here's a blog post about it](https://blog.heroku.com/neither-self-nor-this-receivers-in-go).  I disagree with the section on "Advanced Naming Technique" and actually the linter will flag that kind of thing nowadays.

A more idiomatic name for the receiver here would be `js`, `jpegScrubber`, or `scrubber`.

File: scrubber.go (hunk 6 above)
Comment (hwh33) [reply scrubber.go:23]: Also, it might be helpful to add context to this error, like `fmt.Errorf("unable to read content type: %w", err)`.  Might be good to reset the reader in that case too.  That way the caller can proceed to use the non-scrubbing reader if that's okay for their use case.

File: jpeg/jpeg.go [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+	segmentHeaderBuffer := new(bytes.Buffer)
+	teedReader := io.TeeReader(self.reader, segmentHeaderBuffer)
+
+	reset := func() {
+		self.reader = io.MultiReader(segmentHeaderBuffer, self.reader)
+	}
+
+	segmentMarker := make([]byte, 2)
+	if _, err = io.ReadFull(teedReader, segmentMarker); err != nil {
+		reset()
```
Comment (hwh33) [inline jpeg/jpeg.go]: Why is it necessary to reset the reader here?  This reflects some kind of corruption in the file structure, right?  If `self.reader` is not returning well-formed JPEG data, it seems okay to me to just return an error.

File: jpeg/jpeg.go [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+	// 2 byte segmentDataLength + segmentDataLength (doesn't count marker bytes)
+	segmentLength := int64(2 + segmentDataLength)
+
+	segmentReader := io.MultiReader(segmentHeaderBuffer, self.reader)
+
+	if isMetadataType(segmentMarker) {
+		if _, err = io.CopyN(ioutil.Discard, segmentReader, segmentLength); err != nil {
+			// TODO: some kind of reset here?
+			return
```
Comment (hwh33) [inline jpeg/jpeg.go]: It would be helpful to add context to the error here.  Something like `return n, fmt.Errorf("unable to read segment: %w", err)`

File: jpeg/jpeg.go [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+		n, err = io.ReadFull(self.reader, buf[:len(headerBytes)])
+		if err != nil {
+			return
+		}
+		self.wroteHeader = true
+	}
+
+	n2, err = self.segmentData.Read(buf[n:])
+	n += n2
+	if err != io.EOF {
```
Comment (hwh33) [inline jpeg/jpeg.go]: I think you want to check for non-nil and non-EOF errors.  `self.segmentData` reads from `self.reader` which can return who-knows-what errors.

File: jpeg/jpeg.go (hunk 12 above)
Comment (hwh33) [reply jpeg/jpeg.go]: This goes for all of the reset logic by the way.  I'm not sure it helps the caller out at all and it adds complexity here.

File: jpeg/jpeg.go [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+		reset()
+		return
+	}
+
+	// TODO: apparently there are some number of static-length markers????
+	// https://github.com/dsoprea/go-jpeg-image-structure/blob/d40a386309d24fb714c60dcf1f5f88bff3ad9237/splitter.go#L219
+	var segmentDataLength uint16
+	if err = binary.Read(teedReader, binary.BigEndian, &segmentDataLength); err != nil {
+		reset()
+		return
```
Comment (hwh33) [inline jpeg/jpeg.go]: Context would be helpful on this error too.

Comment (hwh33) [reply jpeg/jpeg.go]: If you really wanted, you could have a type called MalformedDataError and return errors of that type in cases like this.  Then callers can use `errors.As` to check whether the error is data corruption or something else (e.g. an I/O error).

File: png/png.go [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+	if err != io.EOF {
+		return
+	}
+
+	chunkHeaderBuffer := new(bytes.Buffer)
+	teedReader := io.TeeReader(self.reader, chunkHeaderBuffer)
+
+	reset := func() {
+		self.reader = io.MultiReader(chunkHeaderBuffer, self.reader)
+	}
```
Comment (hwh33) [inline png/png.go]: Same comment on the reset logic here as in `jpeg`.

**Review by hwh33 — CHANGES_REQUESTED** [review]
Nice work @max-b!  I'm done with my first round of review.

As you noted, there's quite a bit of overlap between the `jpeg` and `png` packages.  If you're looking to address that, an internal interface like this might help:
```
type segmentReader interface {
	// Returns io.EOF when no more segments exist. Calls to nextSegment may invalidate previously-retrieved segments.
	nextSegment() (r io.Reader, isMetadata bool, err error)
}
```
(the caveat about invalidating previous segments might be necessary to allow this to remain a streaming solution)

Then you could unify the scrubbing with something like:
```
type metaScrubber struct {
	sr segmentReader
}

func (ms *metaScrubber) Read(p []byte) (n int, err error) {
	var (
		m      int
		s      io.Reader
		isMeta bool
	)
	for s, isMeta, err = ms.sr.nextSegment(); err == nil; s, isMeta, err = ms.sr.nextSegment() {
		if isMeta {
			continue
		}
		m, err = s.Read(p[n:])
		n += m
		if err != nil || n >= len(p) {
			return
		}
	}
	return
}
```

I would suggest also thinking about just moving the JPEG and PNG logic into the top-level meta-scrubber package.  If you would like public API specific to scrubbing those filetypes, you could have functions like `metascrubber.NewJPEGScrubber(io.Reader) (io.Reader, error)`.

File: cli/main.go (hunk 1 above)
Comment (anacrolix) [reply cli/main.go]: I find the golints to be pedantic and annoying for the most part, fwiw.

**Comment by anacrolix** [comment]
I'd mostly echo @hwh33 's comments. Did you find this approach worthwhile, or was mat2 still an option?

**Review by anacrolix — COMMENTED** [review]
^

File: scrubber_test.go (hunk 7 above)
Comment (max-b) [reply scrubber_test.go]: Rad thanks! I *really* appreciate all this go idiom help!!

File: scrubber_test.go (hunk 9 above)
Comment (max-b) [reply scrubber_test.go:79]: Agreed

File: jpeg/jpeg.go (hunk 11 above)
Comment (max-b) [reply jpeg/jpeg.go]: arghhhh I am struggling to get used to this idiom, but thanks for pointing it out. I think you're right to point this out and I think in general it's probably better for me to get on board with this for the name of consistency.

File: jpeg/jpeg.go (hunk 12 above)
Comment (max-b) [reply jpeg/jpeg.go]: > Why is it necessary to reset the reader here? This reflects some kind of corruption in the file structure, right? If self.reader is not returning well-formed JPEG data, it seems okay to me to just return an error.

I think this would generally represent either corruption in the file structure OR an error reading from the upstream reader.

I'm still learning some go practices, but the thing I was worried about was returning the correct `n` bytes read value vis a vis the amount that the reader has been advanced. Is it "mandatory" to discard a reader after an error has been returned from it or is there some expectation of recovery? If so does `io.ReadFull` or `binary.Read` advance the reader? Am I over-thinking this???

File: jpeg/jpeg.go (hunk 13 above)
Comment (max-b) [reply jpeg/jpeg.go:79]: Does this kind of context wrapping maintain the error type vis a vis `errors.Is`?

File: jpeg/jpeg.go (hunk 14 above)
Comment (max-b) [reply jpeg/jpeg.go]: Huh, I thought that's what I was doing? Basically the only acceptable state for `err` to be in after line 45 is `io.EOF`. I guess if `err` is `nil` there's a fairly serious logic flaw in my code.... I wonder what the correct way to represent that would be...

**Comment by max-b** [comment]
Thanks so much for the thorough review @hwh33 !! I **_really_** appreciate all the go specific/idiom help. I'm working through these and they're suuuper helpful.

**Comment by max-b** [comment]
> I'd mostly echo @hwh33 's comments. Did you find this approach worthwhile, or was mat2 still an option?

I _thiiink_ for our use case, this is a better option. `mat2` can basically only be run on a server (it has a million dependencies), so unless we move to the upload server approach, it's probably not an option.

The other go-only client side options only supported a couple file types anyways (I'd need to review, but I think `png`, `jpeg`, and `tiff`), so if I can get this library to a place where it reliably supports those I think it'll work better.

What might be the next step (after incorporating PR feedback and cleaning up a couple TODOs) is some sort of semi-automated corpus testing where I could get a large number of files, run them through the scrubber, and then validate (in some way or another) that they remain viewable. We can see where that lines up re: priorities.

File: jpeg/jpeg.go (hunk 15 above)
Comment (max-b) [reply jpeg/jpeg.go]: Ooooooh interesting. I think that makes sense...

File: scrubber.go (hunk 6 above)
Comment (max-b) [reply scrubber.go:23]: I was trying to think of what an error attempting to read here would represent. It seems to me that it would be something upstream, like reading from the network or a file or ?? I'm still getting a handle on go idioms, but it seems like I'd want to return the error undiluted??

File: jpeg/jpeg.go (hunk 12 above)
Comment (hwh33) [reply jpeg/jpeg.go]: So I don't think the doc on `io.Reader` is crystal clear on this, but I believe the `n` value returned should be the number of bytes read _into the input byte slice_.  That is, if you read 20 bytes off the underlying reader and discard 5, you should return n = 15.

For example, think about a `tls.Conn`, which [reads off an underlying transport connection](https://golang.org/pkg/crypto/tls/#Client).  When I ask to read from the `tls.Conn`, I only want to know how many bytes were read into my byte slice.  I don't care that TCP flags were read, I just want the TLS payload.

I don't believe there is any expectation of recovery.  IMO, there's a bit of a contract in the API: you give me a JPEG reader, I'll scrub the metadata for you.  If a caller gives you a reader to a corrupt JPEG structure, the contract is violated and the best I can do for you is tell you so.  Same goes for other errors in the underlying reader.  I _think_ this is generally the case in Go API, but this is also kind of my opinion.  I think you're free to define recovery guarantees and strategies in your API.  However, I would shy away from that unless there was a strong and common use case you're trying to support.  

To add to that last point a little - since the caller gets to supply the underlying reader, they can always supply something that's going to buffer what's read and recover on error as they deem appropriate.  So callers aren't left high-and-dry if they want this kind of thing.

File: jpeg/jpeg.go (hunk 14 above)
Comment (hwh33) [reply jpeg/jpeg.go]: Whoops, yeah you're right!

Actually, you can get a `nil` error here and you _do_ want to return in that case.  It means that there is still data left to read in the segment, but you've filled the buffer.  TBH, I was thinking the same thing as you, but I put some print statements in, ran the tests, and it made sense after that.

Anyway, sorry for the confusion, this is correct as-is =)

File: scrubber.go (hunk 6 above)
Comment (hwh33) [reply scrubber.go:23]: Actually, I find that it's a bit rare that it makes sense to return an error undiluted.  The big difference between Go and other languages here is that it's not necessarily easy to get the whole stack trace of the error.  I think attaching context is a way to achieve a similar outcome as it gives you some idea of the call stack that lead to the error.  Don't get me wrong, it definitely seems worse than just normal error-tracing, but it's what we're stuck with.

The [new Go errors API](https://blog.golang.org/go1.13-errors) should help a bit.  It also further incentivizes error-wrapping since callers can use `errors.As` and `errors.Is` to search for possible causes in the call stack.

Thinking about this specific case a bit more, it may make sense to attach the package name to the error here.  TBH, I don't see that too much outside of the standard library, but it seems conceivable that someone could be given a reader from this package and have little to no knowledge of where it came from.  To them, this just represents JPEG data.  If they get an error like `upload failed: read failed: metascrubber: invalid JPEG structure`, it gives them some idea of where to look.

File: scrubber_test.go (hunk 7 above)
Comment (hwh33) [reply scrubber_test.go]: No problem!

**Comment by hwh33** [comment]
No problem - I'm glad to hear it's useful!  This is going to be a neat package

File: jpeg/jpeg.go (hunk 12 above)
Comment (max-b) [reply jpeg/jpeg.go]: Ok yeah I think that makes sense!

File: jpeg/jpeg.go (hunk 10 above)
Comment (max-b) [reply jpeg/jpeg.go]: I might leave it just for documentation-ish purposes

**Comment by max-b** [comment]
@hwh33 I took your advice about refactoring into a `metaScrubber` with a `segmentReader` interface. I think it ends up being a nice abstraction!

I've improved the error handling as well - let me know what you think/whether there's enough context now. I guess I'm still not 100% sure whether there should be additional wrapping inside the `metaScrubber.Read` method...


Thanks!!!

File: scrubber.go [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+	Err     error
+}
+
+func (e *MalformedDataError) Unwrap() error { return e.Err }
+func (e *MalformedDataError) Error() string { return e.Message + ": " + e.Err.Error() }
+
+type segmentReader interface {
+	// Returns io.EOF when no more segments exist.
+	// If isMetadata, r will be an empty reader
+	// The returned reader MUST be exhausted/read until EOF for further calls to nextSegment to be valid
```
Comment (hwh33) [inline scrubber.go:66]: It looks like you violate this requirement in `metaScrubber.Read` by calling `continue` for metadata segments.  It works currently because the implementations pre-discard all data in the reader for metadata segments.

I can see why you have this requirement though.  Maybe discard the metadata in `metaScrubber.Read`?  I think that'd be a little cleaner and it's another thing the implementations don't have to do themselves.

**Review by hwh33 — APPROVED** [review]
Nice @max-b, this looks great!  I'm glad you found the abstraction to be useful.  I left a minor comment about it, but this looks good to me otherwise.

I think the error wrapping is good as-is now.  The segmentReader implementations add enough context, so it doesn't seem necessary for `metaScrubber` to add any more.

File: scrubber.go (hunk 17 above)
Comment (max-b) [reply scrubber.go:66]: Great ok yup that makes sense :+1:

**Comment by max-b** [comment]
Ok cool I'm going to merge this and then I think the next steps will be:
- Add additional file formats (I think pdf is probably highest priority?)
- Incorporate an automated test suite that can use a larger corpus and check that we aren't breaking images in any corner cases. I think that https://github.com/ianare/exif-samples might be an interesting place to start.

Thanks @hwh33 and @anacrolix !!

**Comment by hwh33** [comment]
+1 to PDFs next.  Also forgot to say so earlier, but automated testing against a large corpus sounds like a great idea.

Nice work!
